// @solana/plugin-core/dist/index.browser.cjs
'use strict';

// src/client.ts
var createEmptyClient = () => createClient();
function createClient(value) {
  return addUse(value ?? {});
}
function addUse(value) {
  return Object.freeze(
    Object.defineProperties(
      {},
      {
        ...Object.getOwnPropertyDescriptors(value),
        use: {
          configurable: false,
          enumerable: true,
          value: function(plugin) {
            const result = plugin(value);
            return result instanceof Promise ? createAsyncClient(result) : addUse(result);
          },
          writable: false
        }
      }
    )
  );
}
function createAsyncClient(promise) {
  return Object.freeze({
    catch(onrejected) {
      return promise.then((v) => addUse(v)).catch(onrejected);
    },
    finally(onfinally) {
      return promise.then((v) => addUse(v)).finally(onfinally);
    },
    then(onfulfilled, onrejected) {
      return promise.then((v) => addUse(v)).then(onfulfilled, onrejected);
    },
    use(plugin) {
      return createAsyncClient(promise.then(plugin));
    }
  });
}
function extendClient(client, additions) {
  const result = Object.defineProperties({}, Object.getOwnPropertyDescriptors(client));
  Object.defineProperties(result, Object.getOwnPropertyDescriptors(additions));
  return Object.freeze(result);
}
function withCleanup(client, cleanup) {
  if (DISPOSABLE_STACK_PROPERTY in client) {
    return addCleanupToClientWithExistingStack(
      client,
      cleanup
    );
  } else {
    return addCleanupToClientWithoutExistingStack(client, cleanup);
  }
}
var DISPOSABLE_STACK_PROPERTY = "__PRIVATE__DISPOSABLE_STACK";
function addCleanupToClientWithExistingStack(client, cleanup) {
  client[DISPOSABLE_STACK_PROPERTY].defer(cleanup);
  return client;
}
function addCleanupToClientWithoutExistingStack(client, cleanup) {
  const stack = new DisposableStack();
  if (Symbol.dispose in client) {
    const existingDispose = client[Symbol.dispose];
    stack.defer(() => existingDispose.call(client));
  }
  stack.defer(cleanup);
  const additions = {
    [DISPOSABLE_STACK_PROPERTY]: stack,
    [Symbol.dispose]() {
      stack[Symbol.dispose]();
    }
  };
  return extendClient(client, additions);
}

exports.createClient = createClient;
exports.createEmptyClient = createEmptyClient;
exports.extendClient = extendClient;
exports.withCleanup = withCleanup;
